File: ppc/pafs/solve.py
import sys

mod = 10 ** 9 + 7

def read_and_split():
	ret = []
	for i in input().split():
		ret.append(int(i))
	return ret

n = read_and_split()[0]

graf = []

for i in range(n):
	graf.append([])

for i in range(1, n):
	a, b = read_and_split()
	a -= 1
	b -= 1
	graf[a].append(b)
	graf[b].append(a)

def choose(d, at_least):
	global mod
	ret = 0
	x = 1
	if at_least == 0:
		ret = 1
	for i in range(1, d + 1):
		x = (x * (d + 1 - i)) % mod
		if i >= at_least:
			ret += x
	return ret % mod

wyn = n - 1
for i in range(n):
	wyn += choose(len(graf[i]), 2)
wyn %= mod

def dfs(v, oj):
	global graf, wyn, mod
	d = len(graf[v])
	ret = choose(d - 1, 1)
	mul = choose(d - 2, 0)
	for i in graf[v]:
		if i == oj:
			continue
		take = dfs(i, v)
		wyn = (wyn + 2 * ret * take) % mod
		ret = (ret + mul * take) % mod
	return ret

sys.setrecursionlimit(10 ** 6)
dfs(0, -1)

print(wyn)

File: ppc/pafs/solve.cpp
#include <bits/stdc++.h>
using namespace std;

using ll=long long;
using pii=pair<int,int>;
using pll=pair<ll,ll>;
using vi=vector<int>;
using vll=vector<ll>;
const int nax=1000*1007;
const ll mod=1000*1000*1000+7;

int n;

vi graf[nax];

ll wyn;

ll choose(int d, int at_least)
{
	ll ret=0;
	ll x=1;
	if (!at_least)
		ret=1;
	for (int i=1; i<=d; i++)
	{
		x=(x*(d+1-i))%mod;
		if (i>=at_least)
			ret=(ret+x)%mod;
	}
	return ret;
}

ll dfs1(int v, int oj)
{
	int d=graf[v].size();
	ll ret=choose(d-1, 1);
	ll mul=choose(d-2, 0);
	for (int i : graf[v])
	{
		if (i==oj)
			continue;
		ll wez=dfs1(i, v);
		wyn=(wyn+2*ret*wez)%mod;
		ret=(ret+mul*wez)%mod;
	}
	return ret;
}

int main()
{
	scanf("%d", &n);
	for (int i=1; i<n; i++)
	{
		int a, b;
		scanf("%d%d", &a, &b);
		graf[a].push_back(b);
		graf[b].push_back(a);
	}
	wyn=n-1;
	for (int i=1; i<=n; i++)
		wyn=(wyn+choose(graf[i].size(), 2))%mod;
	dfs1(1, 0);
	printf("%lld\n", wyn);
	return 0;
}
